<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.19"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>metric: /home/joshua/projects/metric/population_genetics_models/DSE.cpp File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">metric
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.19 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_4248d6557d493447b381d8ca08136f3b.html">population_genetics_models</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#namespaces">Namespaces</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">DSE.cpp File Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Contains functions to run the Diploid Single Environment model @detail This model explores biological function in the context of a trait whose spread is influenced, not only by its own effects, but also by the effects of another trait. The set-up is a Wright-Fisher diploid model: one locus, two alleles, (and implictly a single environment). The resident population has trait (genotype) aa, which is "invaded" by our allele of interest, A (i.e. an aa individual mutates to Aa). Let's say that our trait of interest is the heterozygote Aa. Whether A allele persists in the population depends not only on the fitness of Aa but also on the fitness of AA Since I use allele frequencies&mdash;not genotype frequencies&mdash;to calculate the function metric, the function of allele A needs to be apportioned over the traits (genotypes) Aa and AA (as well as their redundant/synergistic effects). For this I use PID, where the two "sources" are the fitnesses of Aa and AA. By combining the function metric and the PID, we can calculate the function of traits Aa and AA.  
<a href="#details">More...</a></p>
<div class="textblock"><code>#include &lt;cassert&gt;</code><br />
<code>#include &lt;string&gt;</code><br />
<code>#include &lt;vector&gt;</code><br />
<code>#include &lt;numeric&gt;</code><br />
<code>#include &lt;random&gt;</code><br />
<code>#include &quot;<a class="el" href="Parameters_8h_source.html">Parameters.h</a>&quot;</code><br />
<code>#include &quot;DSE.h&quot;</code><br />
<code>#include &quot;rng.h&quot;</code><br />
<code>#include &quot;persistence_probability.h&quot;</code><br />
<code>#include &quot;<a class="el" href="print__results_8h_source.html">print_results.h</a>&quot;</code><br />
<code>#include &quot;<a class="el" href="allele__invasion_8h_source.html">allele_invasion.h</a>&quot;</code><br />
<code>#include &quot;DataContainer.h&quot;</code><br />
</div><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="namespaces"></a>
Namespaces</h2></td></tr>
<tr class="memitem:namespaceDSE"><td class="memItemLeft" align="right" valign="top"> &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceDSE.html">DSE</a></td></tr>
<tr class="memdesc:namespaceDSE"><td class="mdescLeft">&#160;</td><td class="mdescRight">Namespace for Diploid Single Environment. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:acd3ff9e8194a169b1a78f6ba9740aa6c"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="structDSE__Model__Parameters.html">DSE_Model_Parameters</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceDSE.html#acd3ff9e8194a169b1a78f6ba9740aa6c">DSE::parse_parameter_values</a> (int argc, char *argv[])</td></tr>
<tr class="memdesc:acd3ff9e8194a169b1a78f6ba9740aa6c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reads in parameter values from command line into a struct.  <a href="namespaceDSE.html#acd3ff9e8194a169b1a78f6ba9740aa6c">More...</a><br /></td></tr>
<tr class="separator:acd3ff9e8194a169b1a78f6ba9740aa6c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5cea6f450699bacb3fad51cc86345ce"><td class="memItemLeft" align="right" valign="top">const std::vector&lt; double &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceDSE.html#ad5cea6f450699bacb3fad51cc86345ce">DSE::get_fitness_function</a> (const <a class="el" href="structDSE__Model__Parameters.html">DSE_Model_Parameters</a> &amp;parameters)</td></tr>
<tr class="memdesc:ad5cea6f450699bacb3fad51cc86345ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates fitness function.  <a href="namespaceDSE.html#ad5cea6f450699bacb3fad51cc86345ce">More...</a><br /></td></tr>
<tr class="separator:ad5cea6f450699bacb3fad51cc86345ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6fc9a9827b3c6dba9e33035533866bf3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceDSE.html#a6fc9a9827b3c6dba9e33035533866bf3">DSE::calculate_allele_freqs</a> (double &amp;allele_A_freq, const std::vector&lt; double &gt; &amp;fitnesses, const <a class="el" href="structDSE__Model__Parameters.html">DSE_Model_Parameters</a> &amp;parameters, std::mt19937 &amp;rng, int &amp;gen)</td></tr>
<tr class="memdesc:a6fc9a9827b3c6dba9e33035533866bf3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates frequency of the A allele after selection and random mating.  <a href="namespaceDSE.html#a6fc9a9827b3c6dba9e33035533866bf3">More...</a><br /></td></tr>
<tr class="separator:a6fc9a9827b3c6dba9e33035533866bf3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae03e43b11e918694ae4a3c244b65c381"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceDSE.html#ae03e43b11e918694ae4a3c244b65c381">DSE::run_model</a> (int argc, char *argv[])</td></tr>
<tr class="memdesc:ae03e43b11e918694ae4a3c244b65c381"><td class="mdescLeft">&#160;</td><td class="mdescRight">Runs Diploid Single Environment model.  <a href="namespaceDSE.html#ae03e43b11e918694ae4a3c244b65c381">More...</a><br /></td></tr>
<tr class="separator:ae03e43b11e918694ae4a3c244b65c381"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Contains functions to run the Diploid Single Environment model @detail This model explores biological function in the context of a trait whose spread is influenced, not only by its own effects, but also by the effects of another trait. The set-up is a Wright-Fisher diploid model: one locus, two alleles, (and implictly a single environment). The resident population has trait (genotype) aa, which is "invaded" by our allele of interest, A (i.e. an aa individual mutates to Aa). Let's say that our trait of interest is the heterozygote Aa. Whether A allele persists in the population depends not only on the fitness of Aa but also on the fitness of AA Since I use allele frequencies&mdash;not genotype frequencies&mdash;to calculate the function metric, the function of allele A needs to be apportioned over the traits (genotypes) Aa and AA (as well as their redundant/synergistic effects). For this I use PID, where the two "sources" are the fitnesses of Aa and AA. By combining the function metric and the PID, we can calculate the function of traits Aa and AA. </p>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="http://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.8.19
</small></address>
</body>
</html>
